File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/java/io/github/lishangbu/avalon/dataset/service/impl/TypeServiceImplTest.java
package io.github.lishangbu.avalon.dataset.service.impl;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.github.lishangbu.avalon.dataset.entity.Type;
import io.github.lishangbu.avalon.dataset.repository.TypeRepository;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

@ExtendWith(MockitoExtension.class)
class TypeServiceImplTest {
    @Mock private TypeRepository typeRepository;
    @InjectMocks private TypeServiceImpl typeService;

    @Test
    void getPageByCondition_callsRepository() {
        Type type = new Type();
        Pageable pageable = PageRequest.of(0, 5);
        Page<Type> page = new PageImpl<>(List.of(type));
        when(typeRepository.findAll(any(Example.class), eq(pageable))).thenReturn(page);

        Page<Type> result = typeService.getPageByCondition(type, pageable);

        assertSame(page, result);
        ArgumentCaptor<Example> captor = ArgumentCaptor.forClass(Example.class);
        verify(typeRepository).findAll(captor.capture(), eq(pageable));
        assertNotNull(captor.getValue().getProbe());
    }

    @Test
    void listByCondition_callsRepository() {
        Type type = new Type();
        List<Type> expected = List.of(type);
        when(typeRepository.findAll(any(Example.class))).thenReturn(expected);

        List<Type> result = typeService.listByCondition(type);

        assertSame(expected, result);
        verify(typeRepository).findAll(any(Example.class));
    }

    @Test
    void save_usesRepository() {
        Type type = new Type();
        when(typeRepository.save(type)).thenReturn(type);

        Type result = typeService.save(type);

        assertSame(type, result);
        verify(typeRepository).save(type);
    }

    @Test
    void update_usesRepository() {
        Type type = new Type();
        when(typeRepository.save(type)).thenReturn(type);

        Type result = typeService.update(type);

        assertSame(type, result);
        verify(typeRepository).save(type);
    }

    @Test
    void removeById_callsRepository() {
        typeService.removeById(1L);
        verify(typeRepository).deleteById(1L);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-service/src/test/kotlin/io/github/lishangbu/avalon/dataset/service/impl/TypeServiceImplTest.kt
package io.github.lishangbu.avalon.dataset.service.impl

import io.github.lishangbu.avalon.dataset.entity.dto.SaveTypeInput
import io.github.lishangbu.avalon.dataset.entity.dto.TypeSpecification
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateTypeInput
import io.github.lishangbu.avalon.dataset.repository.TypeRepository
import org.babyfish.jimmer.sql.ast.mutation.AssociatedSaveMode
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`
import io.github.lishangbu.avalon.dataset.entity.Type as DatasetType

class TypeServiceImplTest {
    private val repository = mock(TypeRepository::class.java)
    private val service = TypeServiceImpl(repository)

    @Test
    fun listByCondition_callsRepository() {
        val specification = TypeSpecification(id = "1", internalName = "fire", name = "火")
        `when`(repository.findAll(specification)).thenReturn(listOf(typeEntity(1L, "fire", "火")))

        val result = service.listByCondition(specification)

        assertEquals(1, result.size)
        assertEquals("1", result.first().id)
        assertEquals("火", result.first().name)
    }

    @Test
    fun save_usesInsertOnlyMode() {
        `when`(repository.save(any<DatasetType>(), eq(SaveMode.INSERT_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())).thenReturn(typeEntity(1L, "fire", "火"))

        val result = service.save(SaveTypeInput("fire", "火"))

        assertEquals("1", result.id)
        verify(repository).save(any<DatasetType>(), eq(SaveMode.INSERT_ONLY), eq(AssociatedSaveMode.REPLACE), isNull())
    }

    @Test
    fun update_usesUpsertMode() {
        `when`(repository.save(any<DatasetType>(), eq(SaveMode.UPSERT), eq(AssociatedSaveMode.REPLACE), isNull())).thenReturn(typeEntity(1L, "water", "水"))

        val result = service.update(UpdateTypeInput("1", "water", "水"))

        assertEquals("1", result.id)
        verify(repository).save(any<DatasetType>(), eq(SaveMode.UPSERT), eq(AssociatedSaveMode.REPLACE), isNull())
    }

    @Test
    fun removeById_callsRepository() {
        service.removeById(1L)

        verify(repository).deleteById(1L)
    }
}

private fun typeEntity(
    id: Long,
    internalName: String,
    name: String,
): DatasetType =
    DatasetType {
        this.id = id
        this.internalName = internalName
        this.name = name
    }
